text/3532-cfg-attribute-in-tuple-type.md

```diff
@@ -0,0 +1,323 @@
+
+```rust
+for (id, (number, &flag)) in world.query_mut::<(&mut i32, &bool)>() {
+  if flag { *number *= 2; }
+}
+```
+
+Tuples have a number of unique advantages in this paradigm. For one, they avoid boilerplate due to their ability to be anonymously constructed on the fly. Additionally, tuples can be concatenated and joined (e.g. via the [tuple](https://docs.rs/tuple) crate). This allows more advanced ECS libraries and other similar tools to provide support for pre-determined bundles of components, or use tuple nesting to group logic and functionality. One could theoretically define functions like `query_mut1<T0>`, `query_mut2<T0, T1>`, and so on, but the ergonomics of the tuple approach win out in practice.
+
+In this situation, cfg-attributes come into play when building ECS archetypes (a pre-determined collection of components for a type of entity) for different platforms or deployment targets. Say for example that we were creating a multiplayer asteroids game in an Entity Component System. If we wanted to statically define our archetypes at compile-time (as is the case in [gecs](https://docs.rs/gecs)), it might look something like this:
+
+```rust
+type ShipArchetype = EcsArchetype<(
+    TransformComponent,
+    VelocityComponent,
+    PhysicsComponent,
+    ColliderComponent,
+    EngineComponent,
+    HealthComponent,
+    WeaponComponent,
+    EnergyComponent,
+    SpriteComponent,
+    AudioComponent,
+)>;
+```
+
+Since this is a multiplayer game, we may want some components to exist solely on the server or on the client, both for security reasons and also for optimization or performance reasons. The sprite and audio components for example serve no purpose on the server as the server does not render graphics or play audio. In games in other languages, it is common practice to use conditional compilation to avoid putting code in various build targets that serve no purpose, waste resources, or potentially leak information to cheaters. So in this case we will restrict these two components to the `client` feature, like so:
+
+```rust
+type ShipArchetype = EcsArchetype<(
+    TransformComponent,
+    VelocityComponent,
+    PhysicsComponent,
+    ColliderComponent,
+    EngineComponent,
+    HealthComponent,
+    WeaponComponent,
+    EnergyComponent,
+    #[cfg(feature = "client")] SpriteComponent,
+    #[cfg(feature = "client")] AudioComponent,
+)>;
+```
+
+Additionally, we need some components to handle serializing the network state, performing dead reckoning, and sending that information to the client from the server. So we will add a `StateStorageComponent` and a `DeltaCompressionComponent`, and restrict those to the server, since the client does not perform these calculations and we want to avoid giving clients this information to help defeat, or at least delay the arms race against, cheaters.
```

Comment:
```suggestion
Additionally, we need some components to handle serializing the network state, performing dead reckoning, and sending that information to the client from the server. So we will add a `StateStorageComponent` and a `DeltaCompressionComponent`, and restrict those to the server, since the client does not perform these calculations and we want to avoid giving clients this information to help confound, or at least delay, cheaters.
```